tests/tests_deployment/test_keycloak_api.py

```diff
@@ -0,0 +1,1845 @@
+"""Keycloak API endpoint tests."""
+
+import os
+import pathlib
+import uuid
+
+import pytest
+import requests
+
+from _nebari.config import read_configuration
+from nebari.plugins import nebari_plugin_manager
+from tests.tests_deployment import constants
+
+from .keycloak_api_utils import KeycloakAPI, decode_jwt_token
+
+
+def get_nebari_config():
+    config_schema = nebari_plugin_manager.config_schema
+    config_filepath = constants.NEBARI_CONFIG_PATH
+    assert pathlib.Path(config_filepath).exists()
+    config = read_configuration(config_filepath, config_schema)
+    return config
+
+
+@pytest.fixture(scope="session")
+def keycloak_base_url() -> str:
+    """Get the base URL for Keycloak."""
+    config = get_nebari_config()
+
+    keycloak_server_url = os.environ.get(
+        "KEYCLOAK_SERVER_URL", f"https://{config.domain}/auth/"
+    )
+    return keycloak_server_url
+
+
+@pytest.fixture(scope="session")
+def keycloak_username() -> str:
+    """Get the Keycloak admin username."""
+    keycloak_admin_username = os.environ.get("KEYCLOAK_ADMIN_USERNAME", "root")
+    return keycloak_admin_username
+
+
+@pytest.fixture(scope="session")
+def keycloak_password() -> str:
+    """Get the Keycloak admin password."""
+    config = get_nebari_config()
+    keycloak_admin_password = os.environ.get(
+        "KEYCLOAK_ADMIN_PASSWORD",
+        config.security.keycloak.initial_root_password,
+    )
+    return keycloak_admin_password
+
+
+@pytest.fixture(scope="session")
+def keycloak_realm() -> str:
+    """Get the Keycloak realm name."""
+    return os.getenv("KEYCLOAK_REALM", "master")
+
+
+@pytest.fixture(scope="session")
+def keycloak_client_id() -> str:
+    """Get the Keycloak client ID for authentication."""
+    return os.getenv("KEYCLOAK_CLIENT_ID", "admin-cli")
+
+
+@pytest.fixture(scope="session")
+def verify_ssl() -> bool:
+    """Determine if SSL verification should be enabled.
+
+    If KEYCLOAK_VERIFY_SSL is 1 or true, SSL verification is enabled.
+    For all other inputs, SSL verification is disabled (default).
+    """
+    verify = os.environ.get("KEYCLOAK_VERIFY_SSL", "false")
+    return verify.lower() in ("1", "true")
+
+
+@pytest.fixture(scope="session")
+def keycloak_api(
+    keycloak_base_url: str,
+    keycloak_realm: str,
+    keycloak_client_id: str,
+    verify_ssl: bool,
+) -> KeycloakAPI:
+    """Create a KeycloakAPI instance without authentication."""
+    return KeycloakAPI(
+        base_url=keycloak_base_url,
+        realm=keycloak_realm,
+        client_id=keycloak_client_id,
+        verify_ssl=verify_ssl,
+    )
+
+
+@pytest.fixture(scope="session")
+def authenticated_keycloak_api(
+    keycloak_base_url: str,
+    keycloak_username: str,
+    keycloak_password: str,
+    keycloak_realm: str,
+    keycloak_client_id: str,
+    verify_ssl: bool,
+) -> KeycloakAPI:
+    """Create an authenticated KeycloakAPI instance for admin operations."""
+    api = KeycloakAPI(
```

Comment:
What happens if the token expires mid test (say the tests are slow)? Do we have token renewal handled for these tests?